### PR TITLE
Toptica fixes

### PR DESCRIPTION
**Pull request by CatherineH**
This fixes several bugs that were not caught in the first iteration of this driver, renames some of the tests for clarity, and adds firmware/serial number getting functionality.

Changes to the toptica topmode module:

- module now uses multichar \r\n instead of \n with string manipulations
- laser.lock_start,  property checks whether the lock has successfully completed or is in process before returning the start time so that empty dates are not returned.
- laser.first_mode_hop_time and laser.latest_mode_hop_time properties check whether a mode hop has been detected before querying these times so that empty dates are not returned.
- removes the runtime error from laser.is_connected property. If the laser is not connected, it simply returns false.
- adds the ability to query the firmware and serial number of the charm controller (not to be confused with the serial number of the laser)
- fixes the date format of the dates returned by the charm controller.

### CODE REVIEW

**Comment by scasagrande** [comment]
I'm cool with the CI push run failing since the PR run is now passing (and we know its just the linting issues).

Could you add a list of bugs that were fixed to the PR description? That way I can include them in the 0.2.0 release notes.

File: instruments/toptica/topmode.py [hunk 1]
```diff
@@ -12,9 +12,11 @@
 from __future__ import division
 from __future__ import unicode_literals
 
+from re import sub
 from builtins import range
 from enum import IntEnum
 
+
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: Extra whitespace?

File: instruments/toptica/topmode.py [hunk 2]
```diff
@@ -137,11 +139,23 @@ def enable(self):
...
-                raise TypeError(
-                    "Laser emmission must be a boolean, got: {}".format(newval))
+                raise TypeError("Emission status must be a boolean, got: "
+                                "{}".format(type(newval)))
+            if not self.is_connected:
+                return
             self.parent.set(self.name + ":enable-emission", newval)
 
         @property
+        def is_connected(self):
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: What about the case when the laser was initially connected, but then is later disconnected? It should timeout during the `self.serial_number` call. We might not want to raise an exception in that case, but maybe return `False`? I guess it depends what kind of "connected" we are going for here. Physically and/or logically.

File: instruments/toptica/topmode.py [hunk 3]
```diff
@@ -224,7 +239,14 @@ def lock_start(self):
             :return: The datetime of start of mode-locking for specified laser
             :type: `datetime`
             """
-            return ctdate(self.parent.reference(self.name + ":charm:reg:started"))
+            if not self.is_connected:
+                return
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: If the laser is not connected, I don't think that we should be silently exiting this call. Someone's code could depend on this returned datetime value, but if we're just returning `None` then they could have problems down the line when debugging.

File: instruments/toptica/topmode.py [hunk 4]
```diff
@@ -310,7 +342,15 @@ def reference(self, param):
         :return: Response to the reference request
         :rtype: `str`
         """
-        return self.query("(param-ref '{})".format(param))
+        # the toptica topmode termination character is \r\n,
+        # but unfortunately serial communicator only allows for single
+        # character terminators.
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: Note that in the develop branch there is support for multi-character terminators.

File: instruments/toptica/topmode.py [hunk 5]
```diff
@@ -224,7 +239,14 @@ def lock_start(self):
...
             :type: `datetime`
             """
-            return ctdate(self.parent.reference(self.name + ":charm:reg:started"))
+            if not self.is_connected:
+                return
+            response = self.parent.reference(self.name + ":charm:reg:started")
+            # if mode locking has not started yet, the device will respond with an empty
+            # date string. This causes a problem with ctdate.
+            if len(response) < 2:
+                return None
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: ditto

File: instruments/toptica/topmode.py [hunk 6]
```diff
@@ -234,7 +256,12 @@ def first_mode_hop_time(self):
             :return: The datetime of the first mode hop for the specified laser
             :type: `datetime`
             """
-            return ctdate(self.parent.reference(self.name + ":charm:reg:first-mh"))
+            response = self.parent.reference(self.name + ":charm:reg:first-mh")
+            # if the mode has not hopped, the device will respond with an empty date
+            # string. This causes a problem with ctdate.
+            if len(response) < 2:
+                return None
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: ditto

File: instruments/toptica/topmode.py [hunk 7]
```diff
@@ -310,7 +342,15 @@ def reference(self, param):
         :return: Response to the reference request
         :rtype: `str`
         """
-        return self.query("(param-ref '{})".format(param))
+        # the toptica topmode termination character is \r\n,
+        # but unfortunately serial communicator only allows for single
+        # character terminators.
+        # toptica also returns all strings encapsulated by double quotations.
+        response = self.query("(param-ref '{})".format(param))
+        response = sub("\r", "", response)
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: For generic substitutions (without specifying other regex stuff, like beginning/end of string) you can always use the Python `String.replace` method, but I'm cool with this.

File: instruments/toptica/topmode.py [hunk 8]
```diff
@@ -139,29 +136,26 @@ def enable(self):
...
             self.parent.set(self.name + ":enable-emission", newval)
 
         @property
         def is_connected(self):
             """
-            Check whether a laser is connected
-
-            :return: Whether the controller successfully connected to a laser
-            :type: `bool`
+            Check whether a laser is even connected
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: Looks like we lost some of the docstring in the process here

File: instruments/toptica/topmode.py [hunk 9]
```diff
@@ -231,7 +225,8 @@ def mode_hop(self):
             :return: Mode-hop status of the specified laser
             :type: `bool`
             """
-            response = self.parent.reference(self.name + ":charm:reg:mh-occurred")
+            response = self.parent.reference(self.name +
+                                             ":charm:reg:mh-occurred")
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: Was pylint complaining about the length of this line?

File: instruments/toptica/topmode.py [hunk 10]
```diff
@@ -242,13 +237,14 @@ def lock_start(self):
...
             :type: `datetime`
             """
-            if not self.is_connected:
-                return
+            # if mode locking has not started yet, the device will respond with
+            # an empty date string. This causes a problem with ctdate.
+            _corr_stat = self.correction_status
+            if _corr_stat == TopMode.CharmStatus.un_initialized \
+                    or _corr_stat == TopMode.CharmStatus.failure:
+                raise RuntimeError("Laser has not yet successfull locked")
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: spelling

File: instruments/toptica/topmode.py [hunk 11]
```diff
@@ -426,7 +417,7 @@ def firmware(self):
         :return: The firmware version of the charm controller
         :type: `str`
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: Should probably update this to the correct type

File: instruments/toptica/topmode.py [hunk 12]
```diff
@@ -426,7 +417,7 @@ def firmware(self):
...
         """
-        firmware = tuple(map(int, self.reference("fw-ver").split(".")))
+        firmware = [int(i) for i in self.reference("fw-ver").split(".")]
```
Comment (scasagrande) [inline instruments/toptica/topmode.py]: We should return a tuple (firmware versions doesn't need to be mutable). Does this follow the usual X.Y.Z formatting? If you think they might come out with updates that changes this (say, from X.Y.Z to X.Y) we might want to enforce a minimum length here by appending zeros until the length is 3. I think we did that for some of the Qubitekk instruments iirc.

File: instruments/toptica/topmode.py (hunk 10 above)
Comment (CatherineH) [inline instruments/toptica/topmode.py]: What did I misspell?

Comment (scasagrande) [inline instruments/toptica/topmode.py]: `successfull` -> `successfully`

File: instruments/toptica/topmode.py (hunk 9 above)
Comment (CatherineH) [inline instruments/toptica/topmode.py]: It was 83 characters. pep says 79.

Comment (scasagrande) [inline instruments/toptica/topmode.py]: 83 is fine by me, I'm not super hard on the 79 limit

**Pull request merged**